### PR TITLE
Update calico to v3.8.2

### PR DESCRIPTION
**Pull request by zanetworker**
**What this PR does / why we need it**:
This PR updates Calico to v3.8.2 and adds Calico-typha with proportional autoscaling. 

**Special notes for your reviewer**:
The default volume plugin path in Gardener should be changed to ` /var/lib/kubelet/volumeplugins` to circumvent [this issue](https://github.com/projectcalico/calico/issues/2712).

This PR depends on: https://github.com/gardener/gardener/pull/1414 

**Release note**:

```improvement operator
Calico now runs on version 3.8.2 and enables typha with proportional autoscaling.
```

### CODE REVIEW

**Comment by ialidzhikov** [comment]
I guess you mean `v3.8.2`.

**Comment by rfranzke** [comment]
Let's test this on Alicloud first before merging

**Comment by zanetworker** [comment]
@adracus works fine on Alicloud.

File: controllers/networking-calico/charts/internal/calico/templates/calico.yaml [hunk 1]
```diff
@@ -280,8 +330,8 @@ spec:
   # We recommend using Typha if you have more than 50 nodes.  Above 100 nodes it is essential
   # (when using the Kubernetes datastore).  Use one replica for every 100-200 nodes.  In
   # production, we recommend running at least 3 replicas to reduce the impact of rolling upgrade.
-  replicas: 0
-  revisionHistoryLimit: 0
+  replicas: 1
+  revisionHistoryLimit: 2
```
Comment (rfranzke) [inline controllers/networking-calico/charts/internal/calico/templates/calico.yaml]: Why? We always use `0` in Gardener because we never rollback anyways.

File: controllers/networking-calico/charts/internal/calico/templates/calico.yaml [hunk 2]
```diff
@@ -334,43 +384,116 @@ spec:
...
+          name: autoscaler
+          command:
+            - /cpvpa
+            - --target=deployment/calico-typha
+            - --namespace=kube-system
+            - --logtostderr=true
+            - --poll-period-seconds=30
+            - --v=2
+            - --config-file=/etc/config/typha-autoscaler
+          volumeMounts:
```
Comment (rfranzke) [inline controllers/networking-calico/charts/internal/calico/templates/calico.yaml]: Checksum annotation for configmap is missing

File: controllers/networking-calico/charts/internal/calico/templates/config.yaml [hunk 3]
```diff
@@ -50,3 +50,51 @@ data:
...
     }
+---
+kind: ConfigMap
+apiVersion: v1
+metadata:
+  name: calico-typha-horizontal-autoscaler
+  namespace: kube-system
+  labels:
+    kubernetes.io/cluster-service: "true"
+    addonmanager.kubernetes.io/mode: Reconcile
```
Comment (ialidzhikov) [inline controllers/networking-calico/charts/internal/calico/templates/config.yaml]: Do we need this addonmanager label?

File: controllers/networking-calico/charts/internal/calico/templates/config.yaml [hunk 4]
```diff
@@ -50,3 +50,51 @@ data:
...
+    }
+---
+kind: ConfigMap
+apiVersion: v1
+metadata:
+  name: calico-typha-vertical-autoscaler
+  namespace: kube-system
+  labels:
+    kubernetes.io/cluster-service: "true"
+    addonmanager.kubernetes.io/mode: Reconcile
```
Comment (ialidzhikov) [inline controllers/networking-calico/charts/internal/calico/templates/config.yaml]: Do we need this addonmanager label?

File: controllers/networking-calico/charts/internal/calico/templates/config.yaml (hunk 3 above)
Comment (rfranzke) [inline controllers/networking-calico/charts/internal/calico/templates/config.yaml]: Remove this label, it's not needed anymore

File: controllers/networking-calico/charts/internal/calico/templates/config.yaml (hunk 4 above)
Comment (rfranzke) [inline controllers/networking-calico/charts/internal/calico/templates/config.yaml]: Remove this label, it's not needed anymore

File: controllers/networking-calico/charts/internal/calico/templates/psp/calico-typha-horizontal-autoscaler-clusterrole.yaml [hunk 5]
```diff
@@ -0,0 +1,15 @@
+
+apiVersion: {{ include "rbacversion" . }}
+kind: ClusterRole
+metadata:
+  name: garden.sapcloud.io:psp:kube-system:typha-cpha
```
Comment (rfranzke) [inline controllers/networking-calico/charts/internal/calico/templates/psp/calico-typha-horizontal-autoscaler-clusterrole.yaml]: Can we call all the new resources `gardener.cloud:...` instead of `garden.sapcloud.io:...`?

File: controllers/networking-calico/charts/internal/calico/templates/rbac.yaml [hunk 6]
```diff
@@ -208,3 +208,95 @@ subjects:
...
   name: calico-kube-controllers
   namespace: kube-system
+---
+apiVersion: {{ include "rbacversion" . }}
+kind: ClusterRole
+metadata:
+  name: typha-cpha
+  labels:
+    kubernetes.io/cluster-service: "true"
+    addonmanager.kubernetes.io/mode: Reconcile
```
Comment (rfranzke) [inline controllers/networking-calico/charts/internal/calico/templates/rbac.yaml]: Remove this label, it's not needed anymore

File: controllers/networking-calico/charts/internal/calico/templates/rbac.yaml [hunk 7]
```diff
@@ -208,3 +208,95 @@ subjects:
...
+    resources: ["nodes"]
+    verbs: ["list"]
+---
+apiVersion: {{ include "rbacversion" . }}
+kind: ClusterRoleBinding
+metadata:
+  name: typha-cpha
+  labels:
+    kubernetes.io/cluster-service: "true"
+    addonmanager.kubernetes.io/mode: Reconcile
```
Comment (rfranzke) [inline controllers/networking-calico/charts/internal/calico/templates/rbac.yaml]: Remove this label, it's not needed anymore

File: controllers/networking-calico/charts/internal/calico/templates/rbac.yaml [hunk 8]
```diff
@@ -208,3 +208,95 @@ subjects:
...
+    namespace: kube-system
+---
+apiVersion: {{ include "rbacversion" . }}
+kind: Role
+metadata:
+  name: typha-cpha
+  namespace: kube-system
+  labels:
+    kubernetes.io/cluster-service: "true"
+    addonmanager.kubernetes.io/mode: Reconcile
```
Comment (rfranzke) [inline controllers/networking-calico/charts/internal/calico/templates/rbac.yaml]: Remove this label, it's not needed anymore

File: controllers/networking-calico/charts/internal/calico/templates/rbac.yaml [hunk 9]
```diff
@@ -208,3 +208,95 @@ subjects:
...
+    verbs: ["get", "update"]
+---
+apiVersion: {{ include "rbacversion" . }}
+kind: RoleBinding
+metadata:
+  name: typha-cpha
+  namespace: kube-system
+  labels:
+    kubernetes.io/cluster-service: "true"
+    addonmanager.kubernetes.io/mode: Reconcile
```
Comment (rfranzke) [inline controllers/networking-calico/charts/internal/calico/templates/rbac.yaml]: Remove this label, it's not needed anymore

File: controllers/networking-calico/charts/internal/calico/templates/rbac.yaml [hunk 10]
```diff
@@ -208,3 +208,95 @@ subjects:
...
+    name: typha-cpha
+    namespace: kube-system
+---
+apiVersion: {{ include "rbacversion" . }}
+kind: ClusterRole
+metadata:
+  name: typha-cpva
+  labels:
+    kubernetes.io/cluster-service: "true"
+    addonmanager.kubernetes.io/mode: Reconcile
```
Comment (rfranzke) [inline controllers/networking-calico/charts/internal/calico/templates/rbac.yaml]: Remove this label, it's not needed anymore

File: controllers/networking-calico/charts/internal/calico/templates/rbac.yaml [hunk 11]
```diff
@@ -208,3 +208,95 @@ subjects:
...
+    resources: ["deployments"]
+    verbs: ["patch"]
+---
+apiVersion: {{ include "rbacversion" . }}
+kind: ClusterRoleBinding
+metadata:
+  name: typha-cpva
+  labels:
+    kubernetes.io/cluster-service: "true"
+    addonmanager.kubernetes.io/mode: Reconcile
```
Comment (rfranzke) [inline controllers/networking-calico/charts/internal/calico/templates/rbac.yaml]: Remove this label, it's not needed anymore

File: controllers/networking-calico/charts/internal/calico/values.yaml [hunk 12]
```diff
@@ -6,7 +6,10 @@ config:
     type: "host-local"
 #    subnet: "usePodCidr"
 images:
-  calico-node: "quay.io/calico/node:v3.7.2"
-  calico-cni: "quay.io/calico/cni:v3.7.2"
-  calico-typha: "quay.io/calico/typha:v3.7.2"
-  calico-kube-controllers: "quay.io/calico/kube-controllers:v3.7.2"
+  calico-node: "quay.io/calico/node:v3.8.2"
```
Comment (rfranzke) [inline controllers/networking-calico/charts/internal/calico/values.yaml]: Please don't use exact image names + tags here. The images should always come from the image vector.
Use `calico-node: image:tag` (etc.) instead.

File: controllers/networking-calico/charts/networking-calico/values.yaml [hunk 13]
```diff
@@ -9,19 +9,32 @@ resources: {}
   # - name: calico-node
   #   sourceRepository: github.com/projectcalico/calico
   #   repository: quay.io/calico/node
-  #   tag: v3.7.2
+  #   tag: v3.8.2
```
Comment (rfranzke) [inline controllers/networking-calico/charts/networking-calico/values.yaml]: Please remove all the concrete images + image tags here, that's double maintenance that we don't need. The image vector is the source of truth.

File: controllers/networking-calico/charts/internal/calico/templates/config.yaml (hunk 4 above)
Comment (zanetworker) [reply controllers/networking-calico/charts/internal/calico/templates/config.yaml]: No.

File: controllers/networking-calico/charts/internal/calico/values.yaml (hunk 12 above)
Comment (zanetworker) [reply controllers/networking-calico/charts/internal/calico/values.yaml]: the images do come from the ImageVector, I ma just following the practice here. We could of course replace these values with any other but what would you recommend to have here?

File: controllers/networking-calico/charts/networking-calico/values.yaml (hunk 13 above)
Comment (zanetworker) [reply controllers/networking-calico/charts/networking-calico/values.yaml]: you mean the tags? this section was added previously for the imageVectorOverwrite.

File: controllers/networking-calico/charts/internal/calico/values.yaml (hunk 12 above)
Comment (rfranzke) [reply controllers/networking-calico/charts/internal/calico/values.yaml]: Nothing, a place holder, sth like https://github.com/gardener/gardener-extensions/blob/master/controllers/provider-aws/charts/internal/machine-controller-manager/seed/values.yaml#L2.
It's confusing to double-maintain the values here.

File: controllers/networking-calico/charts/networking-calico/values.yaml (hunk 13 above)
Comment (rfranzke) [reply controllers/networking-calico/charts/networking-calico/values.yaml]: I would not have something concrete in this image vector overwrite example. If somebody wants to overwrite it the exact values have to be looked up in the image vector anyways.

File: controllers/networking-calico/charts/internal/calico/templates/calico.yaml [hunk 14]
```diff
@@ -334,43 +384,120 @@ spec:
...
-
 ---
-
+apiVersion: {{ include "deploymentversion" . }}
+kind: Deployment
+metadata:
+  name: calico-typha-horizontal-autoscaler
+  namespace: kube-system
+  annotations:
+    checksum/configmap-calico: {{ include (print $.Template.BasePath "/config.yaml") . | sha256sum }}
```
Comment (rfranzke) [inline controllers/networking-calico/charts/internal/calico/templates/calico.yaml]: Why is this here? No container inside this deployment mounts the configmap. Also, the annotation would have needed to be part of the pod template and not part of the deployment metadata.

File: controllers/networking-calico/charts/internal/calico/templates/calico.yaml [hunk 15]
```diff
@@ -334,43 +384,120 @@ spec:
...
+              cpu: 10m
+      serviceAccountName: typha-cpha
+---
+kind: Deployment
+apiVersion: {{ include "deploymentversion" . }}
+metadata:
+  name: calico-typha-vertical-autoscaler
+  namespace: kube-system
+  annotations:
+    checksum/configmap-calico: {{ include (print $.Template.BasePath "/config.yaml") . | sha256sum }}
```
Comment (rfranzke) [inline controllers/networking-calico/charts/internal/calico/templates/calico.yaml]: The annotation needs to be part of the pod template and not part of the deployment metadata.
Also, I would separate the two configs into dedicated files, because right now the checksum you are computing takes into account both the calico-config as well as the autoscaler config. You don't want the autoscaler to restart if the calico config changes and vice versa.

File: controllers/networking-calico/charts/internal/calico/templates/psp/calico-typha-horizontal-autoscaler-clusterrole.yaml [hunk 16]
```diff
@@ -0,0 +1,15 @@
+
+apiVersion: {{ include "rbacversion" . }}
+kind: ClusterRole
+metadata:
+  name: garden.cloud:psp:kube-system:typha-cpha
```
Comment (rfranzke) [inline controllers/networking-calico/charts/internal/calico/templates/psp/calico-typha-horizontal-autoscaler-clusterrole.yaml]: Replace `garden.cloud` with `gardener.cloud` everywhere.

File: controllers/networking-calico/charts/internal/calico/templates/psp/calico-typha-horizontal-autoscaler-clusterrolebinding.yaml [hunk 17]
```diff
@@ -0,0 +1,13 @@
+apiVersion: {{ include "rbacversion" . }}
+kind: RoleBinding
+metadata:
+  name: garden.sapcloud.io:psp:typha-cpa
```
Comment (rfranzke) [inline controllers/networking-calico/charts/internal/calico/templates/psp/calico-typha-horizontal-autoscaler-clusterrolebinding.yaml]: Still old naming scheme is used here

File: controllers/networking-calico/charts/internal/calico/templates/psp/calico-typha-vertical-autoscaler-clusterrolebinding.yaml [hunk 18]
```diff
@@ -0,0 +1,13 @@
+apiVersion: {{ include "rbacversion" . }}
+kind: RoleBinding
+metadata:
+  name: garden.sapcloud.io:psp:typha-cpva
```
Comment (rfranzke) [inline controllers/networking-calico/charts/internal/calico/templates/psp/calico-typha-vertical-autoscaler-clusterrolebinding.yaml]: Still old naming scheme is used here

Comment (rfranzke) [reply controllers/networking-calico/charts/internal/calico/templates/psp/calico-typha-vertical-autoscaler-clusterrolebinding.yaml]: Still not resolved

File: controllers/networking-calico/charts/internal/calico/templates/psp/calico-typha-horizontal-autoscaler-clusterrolebinding.yaml (hunk 17 above)
Comment (rfranzke) [reply controllers/networking-calico/charts/internal/calico/templates/psp/calico-typha-horizontal-autoscaler-clusterrolebinding.yaml]: Still not resolved.

File: controllers/networking-calico/charts/internal/calico/templates/calico.yaml [hunk 19]
```diff
@@ -334,43 +384,120 @@ spec:
...
+    kubernetes.io/cluster-service: "true"
+spec:
+  replicas: 1
+  selector:
+    matchLabels:
+      k8s-app: calico-typha-autoscaler
+  template:
+    metadata:
+      annotations:
+        checksum/configmap-calico-typha-horizontal-autoscaler: {{ include (print $.Template.BasePath "/typha-horizontal-autoscaling-config.yaml") . | sha256sum }}
```
[the remark this answers is not in the document]
Comment (zanetworker) [reply controllers/networking-calico/charts/internal/calico/templates/calico.yaml]: It is used in the command. 
```
      command:
            - /cluster-proportional-autoscaler
            - --namespace=kube-system
            **- --configmap=calico-typha-horizontal-autoscaler**
            - --target=deployment/calico-typha
            - --logtostderr=true
            - --v=2
```

[the remark this answers is not in the document]
Comment (rfranzke) [reply controllers/networking-calico/charts/internal/calico/templates/calico.yaml]: Okay, so it doesn't mount it on the file system but watches it directly from the kube-apiserver and reloads it dynamically? Then we don't need the checksum annotation.

[the remark this answers is not in the document]
Comment (zanetworker) [reply controllers/networking-calico/charts/internal/calico/templates/calico.yaml]: now sure if it watches or just gets the config map.  Will check, I guess if  it just gets then we need it.

[the remark this answers is not in the document]
Comment (rfranzke) [reply controllers/networking-calico/charts/internal/calico/templates/calico.yaml]: Can you find that out please?

[the remark this answers is not in the document]
Comment (zanetworker) [reply controllers/networking-calico/charts/internal/calico/templates/calico.yaml]: It polls using ticks, so we can remove it.

[the remark this answers is not in the document]
Comment (rfranzke) [reply controllers/networking-calico/charts/internal/calico/templates/calico.yaml]: Great, thank you 👍 !
